File: frontend/src/features/generation/pipeline/runner.ts
import type { Processor, ProcessorDescription } from "./types";

/**
 * Runs an ordered list of processors against a shared context.
 *
 * Each processor checks its own activation condition via `isActive()`.
 * If active, its `execute()` is called — it reads from and writes to the context.
 * Processors run sequentially in the order provided.
 */
export async function runProcessors<TContext>(
  processors: readonly Processor<TContext>[],
  context: TContext,
): Promise<void> {
  for (const processor of processors) {
    if (!processor.isActive(context)) {
      continue;
    }
    await processor.execute(context);
  }
}

/**
 * Describes which processors would run for a given context, without executing them.
 *
 * Returns metadata for every processor in the list, with an `active` flag
 * indicating whether it would run. Useful for debugging, logging, and
 * transparency about the pipeline's behavior for a particular workflow+rules.
 */
export function describeProcessors<TContext>(
  processors: readonly Processor<TContext>[],
  context: TContext,
): ProcessorDescription[] {
  return processors.map((processor) => ({
    name: processor.meta.name,
    description: processor.meta.description,
    reads: processor.meta.reads,
    writes: processor.meta.writes,
    active: processor.isActive(context),
  }));
}

File: backend/services/gen_pipeline/runner.py
from __future__ import annotations

import logging
from typing import Any, Sequence

from services.gen_pipeline.types import Processor, ProcessorDescription

log = logging.getLogger(__name__)


async def run_processors(
    processors: Sequence[Processor],
    ctx: Any,
) -> None:
    """Run an ordered list of processors against a shared context.

    Each processor checks its own activation condition via ``is_active()``.
    If active, its ``execute()`` is called — it reads from and writes to
    the context.  Processors run sequentially in the order provided.
    """
    for processor in processors:
        if not processor.is_active(ctx):
            log.debug(
                "[pipeline] Skipping processor %s (not active)",
                processor.meta.name,
            )
            continue
        log.info("[pipeline] Running processor: %s", processor.meta.name)
        await processor.execute(ctx)


def describe_processors(
    processors: Sequence[Processor],
    ctx: Any,
) -> list[ProcessorDescription]:
    """Describe which processors would run for a given context, without executing them.

    Returns metadata for every processor in the list, with an ``active`` flag
    indicating whether it would run.  Useful for debugging, logging, and
    transparency about the pipeline's behavior for a particular workflow+rules.
    """
    return [
        ProcessorDescription(
            name=processor.meta.name,
            description=processor.meta.description,
            reads=processor.meta.reads,
            writes=processor.meta.writes,
            active=processor.is_active(ctx),
        )
        for processor in processors
    ]
